genie/st_mask_git.py

```diff
@@ -0,0 +1,252 @@
+
+        loss_THW = F.cross_entropy(factored_logits, factored_targets, reduction="none").sum(dim=1)
+        acc_THW = (factored_logits.argmax(dim=1) == factored_targets).all(dim=1)
+
+        # Compute the mean masked error.
+        # Multiply loss values by mask instead of indexing them, more computationally efficient.
+        num_masked_tokens = torch.sum(relevant_mask_THW)
+        relevant_loss = torch.sum(loss_THW * relevant_mask_THW) / num_masked_tokens
+        relevant_acc = torch.sum(acc_THW * relevant_mask_THW).float() / num_masked_tokens
+
+        # only optimize on the masked/noised logits?
+        return relevant_loss, relevant_acc
+
+    def compute_logits(self, x_THW):
+        # x_THW is for z0,...,zT while x_targets is z1,...,zT
+        x_TS = rearrange(x_THW, "B T H W -> B T (H W)")
+        x_TSC = self.token_embed(x_TS)
+
+        # additive embeddings, using the same vocab space
+        x_TSC = self.decoder(x_TSC + self.pos_embed_TSC)
+        x_next_TSC = self.out_x_proj(x_TSC)
+
+        logits_CTHW = rearrange(x_next_TSC, "B T (H W) C -> B C T H W", H=self.h, W=self.w)
+        return logits_CTHW
+
+    def forward(self, input_ids, labels):
+        T, H, W = self.config.T, self.h, self.w
+        x_THW = rearrange(input_ids, "B (T H W) -> B T H W", T=T, H=H, W=W)
+
+        logits_CTHW = self.compute_logits(x_THW)
+
+        labels = rearrange(labels, "B (T H W) -> B T H W", T=T, H=H, W=W)
+
+        # Record the loss over masked tokens only to make it more comparable to LLM baselines
+        relevant_mask = x_THW[:, 1:] == self.mask_token_id  # could also get mask of corrupted tokens by uncommenting line in `get_maskgit_collator`
+        relevant_loss, relevant_acc = self.compute_loss_and_acc(logits_CTHW, labels, relevant_mask)
+
+        return ModelOutput(loss=relevant_loss, acc=relevant_acc, logits=logits_CTHW)
+
+    def init_weights(self):
+        """ Works with and without muP. """
+        std = 0.02
+        for module in self.modules():
+            if isinstance(module, nn.Linear):
+                if hasattr(module.weight, "infshape"):  # muP
+                    mup.normal_(module.weight, mean=0.0, std=std)
+                else:
+                    module.weight.data.normal_(mean=0.0, std=std)
+
+                if module.bias is not None:
+                    module.bias.data.zero_()
+            elif isinstance(module, nn.Embedding):
+                module.weight.data.normal_(mean=0.0, std=std)
+                if module.padding_idx is not None:
+                    module.weight.data[module.padding_idx].zero_()
+
+    def set_mup_shapes(self, rescale_params=False):
+        base_config = self.config.shallow_copy()
+        base_config.num_heads = 8
+        base_config.d_model = 256  # currently hardcoding to this shape
+        base_model = STMaskGIT(base_config)
+
+        mup.set_base_shapes(self, base_model, rescale_params=rescale_params)
+
+    @classmethod
```

Comment:
maybe add `@override`? `from typing import override`

Reply:
Not in Python 3.10 I think